ogre/src/OgreLidarVisual.cc

```diff
@@ -0,0 +1,332 @@
+
+  /// \brief Hitting DynamicLines Object to display
+  public: std::vector<std::shared_ptr<OgreDynamicLines>> rayStrips;
+
+  /// \brief Dead Zone Geometry DynamicLines Object to display
+  public: std::vector<std::shared_ptr<OgreDynamicLines>> deadZoneRayFans;
+
+  /// \brief Lidar Ray DynamicLines Object to display
+  public: std::vector<std::shared_ptr<OgreDynamicLines>> rayLines;
+
+  /// \brief The current lidar data message
+  public: std::vector<double> laserMsg;
+
+  /// \brief True if new message is received
+  public: bool receivedMsg;
+};
+
+using namespace ignition;
+using namespace rendering;
+
+//////////////////////////////////////////////////
+OgreLidarVisual::OgreLidarVisual()
+  : dataPtr(new OgreLidarVisualPrivate)
+{
+}
+
+//////////////////////////////////////////////////
+OgreLidarVisual::~OgreLidarVisual()
+{
+  this->Destroy();
+}
+
+//////////////////////////////////////////////////
+void OgreLidarVisual::PreRender()
+{
+  // no ops
+}
+
+//////////////////////////////////////////////////
+void OgreLidarVisual::Destroy()
+{
+  for (auto ray : this->dataPtr->noHitRayStrips)
+  {
+    ray.get()->Clear();
+    ray.reset();
+  }
+
+  for (auto ray : this->dataPtr->rayStrips)
+  {
+    ray.get()->Clear();
+    ray.reset();
+  }
+
+  for (auto ray : this->dataPtr->rayLines)
+  {
+    ray.get()->Clear();
+    ray.reset();
+  }
+
+  for (auto ray : this->dataPtr->deadZoneRayFans)
+  {
+    ray.get()->Clear();
+    ray.reset();
+  }
+
+  this->ClearPoints();
+}
+
+//////////////////////////////////////////////////
+void OgreLidarVisual::Init()
+{
+  BaseLidarVisual::Init();
+  this->Create();
+}
+
+//////////////////////////////////////////////////
+void OgreLidarVisual::Create()
+{
+  this->ClearPoints();
+  this->dataPtr->receivedMsg = false;
+}
+
+//////////////////////////////////////////////////
+void OgreLidarVisual::ClearPoints()
+{
+  this->dataPtr->noHitRayStrips.clear();
+  this->dataPtr->deadZoneRayFans.clear();
+  this->dataPtr->rayLines.clear();
+  this->dataPtr->rayStrips.clear();
+  this->dataPtr->laserMsg.clear();
+  this->dataPtr->receivedMsg = false;
+}
+
+
+//////////////////////////////////////////////////
+void OgreLidarVisual::SetPoints(std::vector<double> &_msg)
+{
+  this->dataPtr->laserMsg.clear();
+  this->dataPtr->laserMsg = _msg;
+  this->dataPtr->receivedMsg = true;
+}
+
+//////////////////////////////////////////////////
+void OgreLidarVisual::Update()
+{
+  if (!this->dataPtr->receivedMsg || this->dataPtr->laserMsg.size() == 0)
+  {
+    ignerr << "Message not received. Exiting update function"
+           << std::endl;
+    return;
+  }
+
+  this->dataPtr->receivedMsg = false;
+  double horizontalAngle = this->minHorizontalAngle;
+  double verticalAngle = this->minVerticalAngle;
+
+  if (this->horizontalCount > 1)
+  {
+    this->horizontalAngleStep =
+        (this->maxHorizontalAngle - this->minHorizontalAngle) /
+              (this->horizontalCount - 1);
+  }
+  if (this->verticalCount > 1)
```

Comment:
can the vertical and horizontal checks be moved to `Init` or `Create`? We should check this early so we don't end up spamming the console on every update

Reply:
I put them here as the parameters can be changed by the user after the lidarvisual is created, hence allowing the user to set the parameters irrespective of the order... for example, the lidarPoints may be set by the user before setting the Vertical or horizontal RayCount and it will only be evaluated after all the parameters have been fully set, i.e. checking performed within the update function. It would also allow the user to dynamically change the parameters if required.

As of now, the update function only checks for 1) if a new message has been received, and 2) if the number of points in the lidarPoints is consistent with vertical and horizontal raycount.

I think all other checks (from the comment below) can be removed as we changed the API that the user cannot set the angle step size, but can set only the max angle, min angle, and ray count for vertical and horizontal rays. I found that there is a significant chance of a user-side-error while setting step size. It is also a redundant parameter, which may cause behaviour that is undesirable to the user if any single parameter is set incorrectly.

Reply:
Shall I make it a function like this instead? [link](https://github.com/ignitionrobotics/ign-rendering/blob/dfcf7c4a574b6b0e94419b7ba1ab3ebce64055f3/ogre/src/OgreGpuRays.cc#L245)

Reply:
just noticed that you changed the `SetVerticalRayCount` and `SetHorizontalRayCount` functions to make sure these values are always greater than 1. So we shouldn't need the check now right?

+1 on removing the angle step size.

Reply:
I edited the `SetVerticalCount` and `SetHorizontalCount` to ensure the values are non-zero. For the case where the value is 1, we do not need to update the step size. Otherwise, the above expression would give a runtime error. For this specific case, the step size does not need to be used as only one ray will be displayed at `minHorizontalAngle` and `minVerticalAngle`.

For all other cases, this expression would be required to calculate step size.